File: src/services/MailService.ts
/**
 * MailService — orchestrator.
 *
 * Decides at startup whether to use IMAP or POP3 based on the presence of
 * IMAP_HOST in the environment configuration, then delegates all protocol
 * work to the appropriate service class.  Exposes a unified async API so
 * that UI screens never need to know which protocol is active.
 */
import { getImapConfig, getPop3Config, getSmtpConfig } from '../config/env';
import { ImapService } from './ImapService';
import { Pop3Service } from './Pop3Service';
import { SmtpService, type OutboundMessage } from './SmtpService';
import { buildConversations } from './ConversationBuilder';
import type { Message } from '../models/Message';
import type { Conversation } from '../models/Conversation';

export class MailService {
  private email: string;
  private password: string;
  private imap: ImapService | null = null;
  private pop3: Pop3Service | null = null;
  private smtp: SmtpService;
  private pollTimer: ReturnType<typeof setInterval> | null = null;
  private subjectMap = new Map<string, string>();

  constructor(email: string, password: string) {
    this.email = email;
    this.password = password;

    const smtpConfig = getSmtpConfig();
    this.smtp = new SmtpService(smtpConfig);
    this.smtp.setCredentials(email, password);

    const imapConfig = getImapConfig();
    if (imapConfig) {
      this.imap = new ImapService(imapConfig);
      this.imap.setCredentials(email, password);
    } else {
      const pop3Config = getPop3Config();
      this.pop3 = new Pop3Service(pop3Config);
      this.pop3.setCredentials(email, password);
    }
  }

  /** Return true when IMAP is being used; false for POP3. */
  get isImap(): boolean {
    return this.imap !== null;
  }

  /** Connect and authenticate against the incoming mail server. */
  async connect(): Promise<void> {
    if (this.imap) {
      await this.imap.connect();
      await this.imap.login();
      await this.imap.selectInbox();
    } else if (this.pop3) {
      await this.pop3.connect();
      await this.pop3.login();
    }
  }

  /**
   * Fetch all messages and return them as grouped Conversation objects.
   * Stores the subject for each message so ConversationBuilder can name groups.
   */
  async fetchConversations(): Promise<Conversation[]> {
    const messages = await this.fetchMessages();
    return buildConversations(messages, this.email, this.subjectMap);
  }

  private async fetchMessages(): Promise<Message[]> {
    if (this.imap) {
      const uids = await this.imap.fetchAllUids();
      return this.imap.fetchMessages(uids);
    } else if (this.pop3) {
      return this.pop3.fetchAllMessages();
    }
    return [];
  }

  /**
   * Send a message.  The subject is set to the groupName for group threads;
   * for one-to-one chats an empty subject is used so the UI stays clean.
   */
  async sendMessage(to: string[], body: string, groupName?: string): Promise<void> {
    const outbound: OutboundMessage = {
      from: this.email,
      to,
      subject: groupName ?? '',
      body,
    };
    await this.smtp.sendMessage(outbound);
  }

  /**
   * Start real-time updates.
   * For IMAP: uses IDLE.  For POP3: starts a polling interval.
   * `onNewMessages` is called with the current full conversation list.
   */
  async startListening(onNewMessages: (conversations: Conversation[]) => void): Promise<void> {
    if (this.imap) {
      await this.imap.startIdle(async () => {
        const conversations = await this.fetchConversations();
        onNewMessages(conversations);
      });
    } else if (this.pop3) {
      const pop3Config = getPop3Config();
      this.pollTimer = setInterval(async () => {
        try {
          await this.pop3!.connect();
          await this.pop3!.login();
          const conversations = await this.fetchConversations();
          await this.pop3!.disconnect();
          onNewMessages(conversations);
        } catch {
          // Polling errors are non-fatal; the next interval will retry.
        }
      }, pop3Config.pollInterval * 1000);
    }
  }

  /** Stop listening for new messages. */
  async stopListening(): Promise<void> {
    if (this.imap) {
      await this.imap.doneIdle();
    }
    if (this.pollTimer) {
      clearInterval(this.pollTimer);
      this.pollTimer = null;
    }
  }

  /** Mark a message as read on the server (IMAP only; no-op for POP3). */
  async markRead(uid: string): Promise<void> {
    if (this.imap) {
      await this.imap.markRead(uid);
    }
  }

  /** Disconnect from all servers. */
  async disconnect(): Promise<void> {
    await this.stopListening();
    if (this.imap) await this.imap.disconnect();
    if (this.pop3) await this.pop3.disconnect();
  }
}

File: src/email/mail_service.rs
/// MailService — orchestrator.
///
/// Auto-selects IMAP or POP3 based on the presence of `ImapConfig`.
/// Exposes a single API used by the route handlers regardless of which
/// protocol is active underneath.

use anyhow::Result;
use std::collections::HashMap;
use crate::config::AppConfig;
use crate::email::{conversation::build_conversations, imap, pop3, smtp};
use crate::models::{Conversation, Message};

/// Fetch all messages and build conversations.
/// Returns the conversation list and the raw message list (for mark-read).
pub async fn fetch_conversations(
    config: &AppConfig,
    email: &str,
    password: &str,
) -> Result<Vec<Conversation>> {
    let (messages, subject_map) = fetch_messages_raw(config, email, password).await?;
    Ok(build_conversations(messages, email, &subject_map))
}

/// Fetch raw messages + subject map.
pub async fn fetch_messages_raw(
    config: &AppConfig,
    email: &str,
    password: &str,
) -> Result<(Vec<Message>, HashMap<String, String>)> {
    if let Some(imap_cfg) = &config.imap {
        imap::fetch_messages(imap_cfg.clone(), email.to_string(), password.to_string()).await
    } else if let Some(pop3_cfg) = &config.pop3 {
        pop3::fetch_messages(pop3_cfg.clone(), email.to_string(), password.to_string()).await
    } else {
        Ok((vec![], HashMap::new()))
    }
}

/// Send a message via SMTP.
pub async fn send_message(
    config: &AppConfig,
    email: &str,
    password: &str,
    to: Vec<String>,
    body: String,
    group_name: Option<String>,
) -> Result<()> {
    smtp::send_message(
        &config.smtp,
        email,
        password,
        smtp::OutboundMessage {
            from: email.to_string(),
            to,
            subject: group_name.unwrap_or_default(),
            body,
        },
    )
    .await
}

/// Mark a message as read (IMAP only; no-op for POP3).
pub async fn mark_read(
    config: &AppConfig,
    email: &str,
    password: &str,
    uid: &str,
) -> Result<()> {
    if let Some(imap_cfg) = &config.imap {
        imap::mark_read(
            imap_cfg.clone(),
            email.to_string(),
            password.to_string(),
            uid.to_string(),
        )
        .await?;
    }
    Ok(())
}
